libmaxminddb recipe

fcelda opened the pull request:
Specify library name and version:  **libmaxminddb/1.6.0**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/conan-io/conan-center-index?pullRequest=4666) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/conan-io/conan-center-index?pullRequest=4666) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/conan-io/conan-center-index?pullRequest=4666) it.</sub>

mathbunnyru commented on recipes/libmaxminddb/all/conanfile.py:
```diff
@@ -0,0 +1,67 @@
+
+    @property
+    def _source_subfolder(self):
+        return "src"
```
```suggestion
        return "source_subfolder"
```

This is a very common default.

fcelda replied:
All right. Understood.

mathbunnyru commented on recipes/libmaxminddb/all/conanfile.py:
```diff
@@ -0,0 +1,67 @@
+
+    @property
+    def _build_subfolder(self):
+        return "build"
```
```suggestion
        return "build_subfolder"
```

mathbunnyru commented on recipes/libmaxminddb/all/conanfile.py:
```diff
@@ -0,0 +1,67 @@
+import os
+import os.path
```
```suggestion
```

mathbunnyru commented on recipes/libmaxminddb/all/test_package/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+cmake_minimum_required(VERSION 3.1)
+project(PackageTest C)
+
+find_package(PkgConfig)
+pkg_check_modules(LIBMAXMINDDB REQUIRED "libmaxminddb")
```
I haven't seen a lot of examples using pkg_config in test_package, so having one more would be great.

fcelda replied:
I'm happy to hear that. I love package config because the interface is lightweight and yet flexible to integrate with all (most?) build systems.

mathbunnyru commented on recipes/libmaxminddb/all/test_package/conanfile.py:
```diff
@@ -0,0 +1,17 @@
+
+    def test(self):
+        if not tools.cross_building(self):
+            self.run(os.path.join(".", "example"))
```
```suggestion
            self.run(os.path.join(".", "example"), run_environment=True)
```

This will allow using shared lib.

fcelda replied:
OK. That makes sense. Included.

uilianries commented on recipes/libmaxminddb/all/conanfile.py:
```diff
@@ -0,0 +1,77 @@
+
+        cmake = CMake(self)
+        cmake.definitions["BUILD_TESTING"] = "OFF"
+        cmake.configure(source_folder=self._source_subfolder, build_folder=self._build_subfolder)
```
```suggestion
        cmake.configure(build_folder=self._build_subfolder)
```

You have to provide CMake wrapper. Please, read https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md#how-to-provide-a-good-recipe

fcelda replied:
Thank you for the pointers.

The library doesn't work with the CMake wrapper out of the box. The build seems to set `INTERFACE_INCLUDE_DIRECTORIES` to include the source directory incorrectly and therefore the configuration fails. I will have a look at that later.

fcelda replied:
This was indeed an upstream bug. It's fixed now with https://github.com/maxmind/libmaxminddb/pull/259

uilianries commented on recipes/libmaxminddb/all/conanfile.py:
```diff
@@ -0,0 +1,77 @@
+    def package(self):
+        cmake = self._configure_cmake()
+        cmake.install()
+        shutil.rmtree(os.path.join(self.package_folder, "lib", "cmake"))
```
```suggestion
        tools.rmdir(os.path.join(self.package_folder, "lib", "cmake"))
```

uilianries commented on recipes/libmaxminddb/all/conanfile.py:
```diff
@@ -0,0 +1,77 @@
+import os
+import shutil
```
```suggestion
```

uilianries commented on recipes/libmaxminddb/all/test_package/conanfile.py:
```diff
@@ -0,0 +1,20 @@
+
+    def test(self):
+        if not tools.cross_building(self):
+            self.run(os.path.join(".", "test_package"), run_environment=True)
```
```suggestion
            bin_path = os.path.join("bin", "test_package")
            self.run(bin_path, run_environment=True)
```

fcelda replied:
All right. I missed `conan_basic_setup()` in the CMake. This is now fixed.

fcelda commented:
I believe this is ready for a review now. Thank you @mathbunnyru and @uilianries for early feedback!

The CMake support in libmaxminddb is rather new and there were several issues that needed to be fixed. I also do not have access to Windows so the CI was my only feedback to get it working there. I squashed the commits as there was a lot of experimentation involved. 

The 4 patches that are included in the Conan recipe were submitted upstream and are now merged in their development branch.

fcelda commented on recipes/libmaxminddb/all/conandata.yml:
```diff
@@ -0,0 +1,15 @@
+    url: "https://github.com/maxmind/libmaxminddb/releases/download/1.5.2/libmaxminddb-1.5.2.tar.gz"
+    sha256: "5237076d250a5f7c297e331c35a433eeaaf0dc205e070e4db353c9ba10f340a2"
+
+patches:
```
https://github.com/maxmind/libmaxminddb/pull/257
https://github.com/maxmind/libmaxminddb/pull/258
https://github.com/maxmind/libmaxminddb/pull/259
https://github.com/maxmind/libmaxminddb/pull/261

mathbunnyru reviewed (approved):
@fcelda congrats on getting this work 👍 

Could we ask the upstream to release the new version with recent changes?
This is not strictly necessary, but if they will be kind enough to do it, we can have a recipe without any patches, which is always good (I mean, I understand, that these patches will disappear for the next version in the config, but we will have to keep the files to be able to build the current version).
If you're not in a hurry, I would ask to create a new release and wait a couple of days.

At the same time, if they do not release, we can always merge this as is.

fcelda commented:
> Could we ask the upstream to release the new version with recent changes?

I will work on that and report back.

fcelda commented:
Gregory was very kind and the new release is out. I've updated the recipe to 1.6.0, dropped all the patches, and squashed the commits again.

I think this PR is ready, @mathbunnyru.

mathbunnyru reviewed (approved):
Nice!

@fcelda just a note: please, try to refrain from squashing commits in this repo. It makes it much more difficult to review the changes.
Bot will automatically squash all your commits and create one beautiful commit in master after a merge.
And it is always nice to be able to see the history of the changes.
